samples/iolib/src/main/cpp/player/SimpleMultiPlayer.cpp

```diff
@@ -24,201 +24,226 @@
+        if (streamState == StreamState::Disconnected) {
+            __android_log_print(ANDROID_LOG_ERROR, TAG, "  streamState::Disconnected");
+        }
+
+        memset(audioData, 0, static_cast<size_t>(numFrames) * static_cast<size_t>
+        (mParent->mChannelCount) * sizeof(float));
+
+        // OneShotSampleSource* sources = mSampleSources.get();
+        for (int32_t index = 0; index < mParent->mNumSampleBuffers; index++) {
+            if (mParent->mSampleSources[index]->isPlaying()) {
+                mParent->mSampleSources[index]->mixAudio((float *) audioData, mParent->mChannelCount,
+                                                         numFrames);
+            }
+        }
+
+        return DataCallbackResult::Continue;
     }
 
-    memset(audioData, 0, static_cast<size_t>(numFrames) * static_cast<size_t>
-            (mParent->mChannelCount) * sizeof(float));
+    void SimpleMultiPlayer::MyErrorCallback::onErrorAfterClose(AudioStream *oboeStream, Result error) {
+        __android_log_print(ANDROID_LOG_INFO, TAG, "==== onErrorAfterClose() error:%d", error);
 
-    // OneShotSampleSource* sources = mSampleSources.get();
-    for(int32_t index = 0; index < mParent->mNumSampleBuffers; index++) {
-        if (mParent->mSampleSources[index]->isPlaying()) {
-            mParent->mSampleSources[index]->mixAudio((float*)audioData, mParent->mChannelCount,
-                                                     numFrames);
+        mParent->resetAll();
+        if (mParent->openStream(PerformanceMode::None) && mParent->startStream()) {
+            mParent->mOutputReset = true;
         }
     }
 
-    return DataCallbackResult::Continue;
-}
+    void SimpleMultiPlayer::MyPresentationCallback::onPresentationEnded(oboe::AudioStream *oboeStream) {
+        __android_log_print(ANDROID_LOG_INFO, TAG, "==== MyPresentationCallback() called");
+    }
 
-void SimpleMultiPlayer::MyErrorCallback::onErrorAfterClose(AudioStream *oboeStream, Result error) {
-    __android_log_print(ANDROID_LOG_INFO, TAG, "==== onErrorAfterClose() error:%d", error);
+    bool SimpleMultiPlayer::openStream(PerformanceMode performanceMode) {
+        __android_log_print(ANDROID_LOG_INFO, TAG, "openStream()");
+
+        // Use shared_ptr to prevent use of a deleted callback.
+        mDataCallback = std::make_shared<MyDataCallback>(this);
+        mErrorCallback = std::make_shared<MyErrorCallback>(this);
+
+        // Create an audio stream
+        AudioStreamBuilder builder;
+        builder.setChannelCount(mChannelCount);
+        // we will resample source data to device rate, so take default sample rate
+        builder.setDataCallback(mDataCallback);
+
+        builder.setFormat(AudioFormat::Float);
+        builder.setSampleRate(48000);
+
+        builder.setErrorCallback(mErrorCallback);
+        builder.setPresentationCallback(mPresentationCallback);
+        builder.setPerformanceMode(performanceMode);
+        builder.setFramesPerDataCallback(128);
```

Comment:
Can you verify whether DrumThumper works similarly to before?

Reply:
DrumThumper may be OK if it passes an explicit LowLatency parameter.
But third-party apps using this as a library class may be impacted if the default behavior changes!